### ISSUE TITLE
mouse position is wrong in fullscreen on i3wm

### ISSUE THREAD

**Issue by crclark** [ggez/ggez#669]
**Describe the bug**
Using the i3 windows manager in fullscreen (either "True" or "Desktop"), the reported mouse position differs from the true cursor position. The discrepancy is 0 in the upper-left corner, and increases moving towards the lower-right corner. Using Gnome on the same machine, fullscreen works correctly.

**To Reproduce**
Here's some code to reproduce.

```rust
use cgmath;
use ggez;

use ggez::event;
use ggez::graphics::{self, DrawMode};
use ggez::mint;
use ggez::{Context, GameResult};
use std::env;
use std::fs::File;
use std::path;

fn display_cursor_pos(ctx: &mut Context, pos: mint::Point2<f32>) {
    let str = format!("{}, {}", pos.x, pos.y);
    let text = ggez::graphics::Text::new(str);
    graphics::draw(ctx, &text, (cgmath::Point2::new(0.0, 0.0),)).unwrap();
}

struct MainState;

impl MainState {
    fn new(_ctx: &mut Context) -> GameResult<MainState> {
        Ok(MainState {})
    }
}

impl event::EventHandler for MainState {
    fn update(&mut self, _ctx: &mut Context) -> GameResult {
        Ok(())
    }

    fn draw(&mut self, ctx: &mut Context) -> GameResult {
        graphics::clear(ctx, [0.1, 0.2, 0.3, 1.0].into());

        //draw a point where we think the cursor is.
        let mouse_pos = ggez::input::mouse::position(ctx);
        let cursor = graphics::Mesh::new_circle(
            ctx,
            DrawMode::fill(),
            cgmath::Point2::new(0.0, 0.0),
            5.0,
            0.1,
            graphics::BLACK,
        )?;
        graphics::draw(ctx, &cursor, (mouse_pos,))?;

        display_cursor_pos(ctx, mouse_pos);

        graphics::present(ctx)?;
        Ok(())
    }
}

fn load_ggez_config(conf_file_path: &path::PathBuf) -> Option<ggez::conf::Conf> {
    let mut conf_file = File::open(conf_file_path).ok()?;
    ggez::conf::Conf::from_toml_file(&mut conf_file).ok()
}

pub fn main() -> GameResult {
    let resource_dir = if let Ok(manifest_dir) = env::var("CARGO_MANIFEST_DIR") {
        let mut path = path::PathBuf::from(manifest_dir);
        path.push("resources");
        path
    } else {
        path::PathBuf::from("./resources")
    };

    let mut conf_file_path = resource_dir.clone();
    conf_file_path.push("ggez_conf.toml");
    let conf = if let Some(conf) = load_ggez_config(&conf_file_path) {
        conf
    } else {
        //write ggez's default conf file to disk
        let mut file = std::fs::File::create(conf_file_path)?;
        ggez::conf::Conf::default().to_toml_file(&mut file).unwrap();
        ggez::conf::Conf::default()
    };

    let cb = ggez::ContextBuilder::new("badmouse", "").add_resource_path(resource_dir);
    let (ctx, event_loop) = &mut cb.conf(conf.clone()).build()?;

    let state = &mut MainState::new(ctx)?;

    let window_mode = conf.window_mode;
    let screen_coords = ggez::graphics::Rect::new(0.0, 0.0, window_mode.width, window_mode.height);
    ggez::graphics::set_screen_coordinates(ctx, screen_coords).unwrap();
    event::run(ctx, event_loop, state)
}
```
Here's my ggez window_mode config:
```
[window_mode]
width = 3440
height = 1440.0
maximized = false
fullscreen_type = "True"
borderless = false
min_width = 0.0
min_height = 0.0
max_width = 0.0
max_height = 0.0
resizable = false
```

**Expected behavior**
The black dot in the example code should be directly under the mouse cursor, like it is when the code is run under Gnome.

**Screenshots or pasted code**
![Screenshot from 2019-08-04 18-26-37](https://user-images.githubusercontent.com/800945/62433119-f96ece80-b6e7-11e9-90aa-b52e71e2b195.png)

**Hardware and Software:**
 - ggez version: 0.5
 - OS: Ubuntu 18.04
-  i3 version: 4.14.1
 - Graphics card: NVidia GTX 1080
 - Graphics card drivers: NVidia's drivers

**Comment by icefoxen** [comment]
Thanks for the bug report, and sorry it took so long!  I will try to reproduce.  Unfortunately my hunch is that this is an issue with DPI scaling, which is, unfortunately, not a rabbit-hole I'm going to dive down any time soon.  See #587 for the gory details.

Can you double-check whether this happens in windowed mode?  And run the `logging` example and look for a line like `[INFO ][winit::platform::platform::x11::window] Guessed window DPI factor: 2` and copy-paste it here?

**Comment by crclark** [comment]
> Can you double-check whether this happens in windowed mode?

Sorry, I forgot to mention that this doesn't happen in windowed mode.

I'll try to get the `logging` example output to you soon. I assume you want the output for running that example in both windowed and fullscreen.

**Comment by crclark** [comment]
I ran the logging example under i3 and gnome, both windowed and true fullscreen.The DPI factors are indeed different between the two. On i3 (windowed and fullscreen), the DPI factor is 1.1666666666666667. On Gnome (windowed and fullscreen), it's 1.

**Comment by icefoxen** [comment]
Heck, I thought winit stopped bothering with non-integer hidpi factors. Ugh. Well, if we can track down where it is getting that number we can at least tell them about it.

**Comment by fatho** [comment]
I think I just ran into the scaling-factor problem as well. I'm running the i3 window manager without any fully-fledged desktop environment.

It appears that `winit` checks the `Xft.dpi` setting in the xresource database, and if that is not present, it computes the scaling factor based on the reported screen size:
https://github.com/rust-windowing/winit/blob/cd5caf6a2207db7f67cdbdc674e8c61fa5e01918/src/platform/linux/x11/util/randr.rs#L125-L128

That code is also still present in the latest version of winit:
https://github.com/rust-windowing/winit/blob/5a78fe33e84023ca4c89cc30ebeebac0fe89e8f0/src/platform_impl/linux/x11/util/randr.rs#L146-L156

The reason why this problem shows up in i3 in particular might be that other desktop environments set the `Xft.dpi` setting themselves based on what the user chose in the system settings (https://unix.stackexchange.com/a/472168).
After adding `Xft.dpi: 96` to my `~/.Xresources` config manually and reloading it, `winit` now reports the expected scaling factor of 1.

**Comment by icefoxen** [comment]
Try this in the `devel` branch, `winit` has been updated by many versions and has significantly improved high DPI support.

**Comment by icefoxen** [comment]
Yeah, confirmed this is a hidpi issue.  Should be fixed in #884 .
